### PR TITLE
Implement param pointers to Nulecule spec.

### PR DESCRIPTION
**Pull request by cdrage**
This implements XPathing as per the Atomic App PR https://github.com/projectatomic/atomicapp/pull/366.

Fixes issue #70.

I'm not too good at modfying / changing spec information, so will need some input from @bexelbie and @aweiteka

### CODE REVIEW

File: spec/README.md [hunk 1]
```diff
@@ -186,6 +186,37 @@ Field Name | Type | Description
 <a name="parametersDefault"></a>default | `string` | **Optional** An optional default value for the parameter.
 <a name="parametersHidden"></a>hidden | `string` | **Optional** An optional boolean signifying the parameter should be obscured when displayed.
 
+##### XPathing
+XPathing is supported by the Nulecule spec as by [RFC 6902](https://tools.ietf.org/html/rfc6902).
```
Comment (goern) [inline spec/README.md]: its "Nulecule Specification"

File: spec/README.md [hunk 2]
```diff
@@ -186,6 +186,37 @@ Field Name | Type | Description
...
+{
+  "name": "password",
+  "description": "mongoDB Admin password",
+  "hidden": true,
+  "constraints": [
+    {
+      "allowed_pattern": "[A-Z0-9]+",
+      "description": "Must consist of characters and numbers only.",
+      "default": 0,
+      "param": ['/spec/containers/0/param']
```
Comment (goern) [inline spec/README.md]: double quotes?

File: spec/README.md [hunk 3]
```diff
@@ -186,6 +186,37 @@ Field Name | Type | Description
...
+ex.
+```yaml
+description: mongoDB Admin password
+hidden: true
+constraints:
+  - allowed_pattern: "[A-Z0-9]+"
+    description: Must consist of characters and numbers only.
+    default: 0
+    param:
+      - /spec/containers/0/param
```
Comment (goern) [inline spec/README.md:202]: no quotes?

File: spec/README.md (hunk 1 above)
Comment (cdrage) [inline spec/README.md]: Ah, I should clarify that the RFC implements JSON pointers (inherentdly how we do xpathing)

File: spec/README.md (hunk 3 above)
Comment (cdrage) [inline spec/README.md:202]: nope, no quotes needed in yaml

**Comment by goern** [comment]
can you add a little why and what to the XPath chapter. Same for "XPathing", it seems to be called JSON patching, as in the RFC you mentioned.

File: spec/README.md (hunk 2 above)
Comment (cdrage) [inline spec/README.md]: errrr we should be using double quotes, I'll change that.

File: spec/param.json [hunk 4]
```diff
@@ -28,6 +28,10 @@
             "description": "An optional boolean signifying the parameter should be obscured when displayed.",
             "type": "boolean",
             "default": false
-        }
+        },
```
Comment (aweiteka) [inline spec/param.json:31]: In this file we're providing a programatic view of what the data structure looks like, not an example.

File: spec/param.json [hunk 5]
```diff
@@ -28,6 +28,10 @@
...
+        "param": [
```
Comment (aweiteka) [inline spec/param.json]:
```
...
        "param": {
            "description": "An array of XPath strings per RFC 6902.",
            "type": "array",
            "default": null
        },
...
```

File: spec/README.md [hunk 6]
```diff
@@ -186,6 +186,37 @@ Field Name | Type | Description
 <a name="parametersDefault"></a>default | `string` | **Optional** An optional default value for the parameter.
 <a name="parametersHidden"></a>hidden | `string` | **Optional** An optional boolean signifying the parameter should be obscured when displayed.
 
+##### XPathing
```
Comment (aweiteka) [inline spec/README.md]: "XPathing" doesn't seem to be a common, well-known term for this. How about "XPath Reference"?

File: spec/README.md [hunk 7]
```diff
@@ -186,6 +186,37 @@ Field Name | Type | Description
...
+XPathing is supported by the Nulecule specification by implementing [RFC 6902](https://tools.ietf.org/html/rfc6902) JSON pointers.
```
Comment (aweiteka) [inline spec/README.md]: How about... "Parameters may be defined by XPath reference or variable substitution. The Nulecule specification implements and conforms to [RFC 6902](https://tools.ietf.org/html/rfc6902)." Or something like that.

**Comment by aweiteka** [comment]
Per [atomicapp pr#366](https://github.com/projectatomic/atomicapp/pull/366#issue-113960183) shouldn't this be part of the artifacts object?

**Comment by cdrage** [comment]
@aweiteka @goern 

Yes, it's suppose to be part of the `artifacts` object but I was unable to find a clean example to indicate it (the spec uses different definitions than our atomic app implementation).

~~Could one of you please take-over this issue and possibly create another PR to merge in? I'm not too good at modifying/improving specifications.~~ 
let me update and try again :)

File: spec/param.json (hunk 5 above)
Comment (cdrage) [inline spec/param.json]: I'm a tad confused, shouldn't this be an array with [ rather than {?

Since we're passing:

``` yaml
    param:
      - /spec/containers/0/param
```

similar to yaml ^^?

File: spec/README.md (hunk 6 above)
Comment (cdrage) [inline spec/README.md]: Maybe we should remove xpathing all together and use pointers instead since most people would understand that better than xpathing.

File: spec/param.json (hunk 5 above)
Comment (aweiteka) [inline spec/param.json]: > I'm a tad confused, shouldn't this be an array with [ rather than {?

I know it's confusing. We're actually describing the param object using a json obj so we're saying it's an array type.

**Comment by cdrage** [comment]
ping @aweiteka @goern updated again

**Comment by cdrage** [comment]
timeout... pinging again
@goern @aweiteka

and @vpavlin

**Comment by aweiteka** [comment]
@cdrage I'm confused how this relates to this[1]. It seems like a duplication. I don't want to ask more of you so I think at this point I would like to take a pass at this. Would you mind?

[1] https://github.com/projectatomic/nulecule/commit/7a83d58c7ff34dbc00c94be252921d384e6092b0

**Comment by cdrage** [comment]
@aweiteka 
That would be my mistake, seems I had merged xpathing during the 0.3.0 release by mistake. I've since reverted that.

Feel free to check this PR again :)

**Comment by aweiteka** [comment]
> ```
> param:
>   - /spec/containers/0/param
> ```

I still can't reconcile this with atomicapp implementation https://github.com/projectatomic/atomicapp/pull/366 . I don't think it belongs in constraints. I also believe it is better described in artifacts, not params definition.

**Comment by aweiteka** [comment]
@cdrage I submitted PR to update your branch.

**Comment by cdrage** [comment]
ping @aweiteka @bexelbie can we merge this?

**Comment by cdrage** [comment]
timeout, pinging again @aweiteka @bexelbie

**Comment by goern** [comment]
LGTM 

@vpavlin ?

**Comment by bexelbie** [comment]
I agree with @aweiteka that this belongs in the artifacts section or at least not in constraints. 

If atomic app varies from the spec we should rationalize the two. Perhaps we need a working spec call?

**Comment by cdrage** [comment]
@bexelbie @goern @dustymabe 

This is badddd documentation in my regards, this should have never been committed. I will update this. 

I also vote to move more towards `yaml` instead since it gives a lot more clarity in regards to the spec (not only the xpathing examples).

This is what xpathing looks like within Atomic App and thus within `artifacts`:

``` yaml
artifacts:
   kubernetes:
       - resource: artifacts/kubernetes/template.json
         params:
            foo:
               - /spec/template/metadata/foo
            bar:
               - /spec/template/metadata/bar
      - resource: artifacts/kubernetes/extra.json
```

**Comment by dustymabe** [comment]
@cdrage.. can we get a TL;DR for this? I'm not sure I follow everything.

**Comment by dustymabe** [comment]
ping @cdrage

**Comment by cdrage** [comment]
TLDR:

If you wish to change a value within a json/yaml file, you can specify the value and position in the file to replace, ex:

```
            param1:
                - /spec/containers/0/ports/0/hostPort
                - /spec/containers/0/ports/0/hostPort2
```

**Comment by cdrage** [comment]
@goern @aweiteka 

Once #209 is merged, I can add this to the spec as it'll be easier to understand from a usability perspective on how to use this.
